# extract consensus sequence of Tripsacum to be treated as outgroup
angsd -doFasta 2 -doCounts 1 -i Tripsacum.bam -out Tripsacum
#-checkBamHeaders 0 -rf scaffoldNamesWithData.txt 
#Those two flags are important, as in Tripsacum some small scaffolds have no reads mapped on. In that case, the consensus sequence file of Tripsacum (Tripsacum.fa) will have no sequence for those scaffolds. 
#While those scaffold exist in the three ingroup bam files, angsd will scream "chromosome length inconsistency error". 
#-checkBamHeaders 0 just ask it to skip checking the bam headers. The analyzed chromosomes are given by "-rf scaffoldNamesWithData.txt", which is generated by "cut -f1 Tripsacum.fa.fai" (the first column of samtools faidxed Tripsacum.fa file)
#-bam P3_luxurians.txt simply gives the directory of the three ingroup bam files
angsd -doAbbababa 1 -blockSize 1000 -anc Tripsacum.fa -doCounts 1 -bam P3_luxurians.txt -uniqueOnly 1 -minMapQ 30 -minQ 20 -minInd 3 -P 30 -checkBamHeaders 0 -rf scaffoldNamesWithData.txt -out luxurians

#jackKnife.R is provided with angsd.
Rscript jackKnife.R file=luxurians.abbababa indNames=P3_luxurians.txt outfile=luxurians

#This generates results as following:
H1      H2      H3      nABBA   nBABA   Dstat   jackEst SE      Z       
RIMMA0703_Mex_Lowland.IndelRealigned.bam        luxurians.bam   RIMMA1007_Gua_Highland.IndelRealigned.bam       75433   352536  -0.6474838      -0.6474838      0.00143925      -449.8758       
RIMMA1007_Gua_Highland.IndelRealigned.bam       luxurians.bam   RIMMA0703_Mex_Lowland.IndelRealigned.bam        76048   352536  -0.6451197      -0.6451197      0.001447166     -445.7814       
RIMMA1007_Gua_Highland.IndelRealigned.bam       RIMMA0703_Mex_Lowland.IndelRealigned.bam        luxurians.bam   76048   75433   0.004059915     0.004059915     0.002820145     1.439612        
#the more detailed explanation can be found at http://www.popgen.dk/angsd/index.php/Abbababa.
